### ISSUE TITLE
"Blaze Helper Context", event handler 2nd param, and Template.instance()

### ISSUE THREAD

**Issue by faceyspacey** [timbrandin/blaze-react#27]
I'm assuming by "Blaze Helper Context" in the checklist you're referring to these very things. @TwinTailsX happened to have just been working on callbacks/autorun/subscribe/etc and they all revolve around this: 

https://github.com/TwinTailsX/sideburns/commit/3ccb4ac269d44f77581c04c8862af70cfb7afb53#diff-ca7c825ad21ebd08da3fae2ace49a2baR53

I noticed he wasn't using the static singleton context of `Template[name]` rather than a true instance context. That actually can be fixed easily for the callbacks and therefore autorun/subscribe. But for the other places mentioned in the title of this post, it's more complex--how do you think we best can do that? What path have you been thinking of? 

It seems the `context` ( https://github.com/timbrandin/sideburns/blob/devel/template.jsx#L63 )  passed into templates has helper methods whose context is the `component` itself ( https://github.com/timbrandin/sideburns/blob/devel/template.jsx#L27 ). But that context really needs to be the true "data" context, which itself is a combination of `this.data` from `getMeteorUserData`, `with` contexts, `each` contexts, and **props** when the context is setup via a template inclusion. Now only after that is addressed, can we get to the "component" context of `Template.instance()` and the 2nd param in event handlers (whose context is the current incorrect context being passed in, that of the react component instance--so addressing that should be a lot easier in fact).

Tell me if I'm wrong about any of these things. It basically seems the stage we are at is we have the helpers being passed along, but their data contexts isn't being properly set, which is a challenging job since it can come from so many places. It seems to me one possible path is implementing a `lookup()` method much like Blaze which checks all these places, and we have a regex replace that injects just `RT.lookup()`, thereby achieving what seems your goal of keeping the actual transpilation to a minimum (so we dont suffer death by brittle regexes).

**Comment by timbrandin** [comment]
I've merged @TwinTailsX's PR and solved what you mentioned! :+1:

Regarding Template.instance() and this in helpers and events I'm looking into it right now.
I'll write an update here after my investigations.

**Comment by timbrandin** [comment]
I did some research into how one could solve the issue of correct `this` in helpers and events and created a solution for it https://github.com/timbrandin/sideburns/blob/devel/template.jsx#L24-L30. 

Though I had to do a workaround for helpers within an each-in helper to get the correct context.
https://github.com/timbrandin/react-templates/blob/master/react-regex.js#L60
https://github.com/timbrandin/react-templates/blob/master/exports.js#L31-L41

I've also attached the component as the last argument to all helpers, so it should be easy to start working with the instance, I guess we could do a something that grabs that instance using `Template.instance()` or so.

And now I've renamed the old check function to be called lookup, I hope it's more clear what it does now.

**Comment by faceyspacey** [comment]
Nice progress. There's a lot of scenarios we're gonna have to support. Here's how Meteor does it:

https://github.com/meteor/meteor/blob/dc3cd6eb92f2bdd1bb44000cdd6abd1e5d0285b1/packages/blaze/lookup.js#L133

And here are some modifications @mitar made for his Blaze Components package:

https://github.com/peerlibrary/meteor-blaze-components/blob/master/lookup.js#L94

His modifications give you an idea of the sorta things we can do. 

We could take 2 approaches here: one by one add support for more scenarios, or take a step back, come up with a plan, refactor some things, and then abstract all the necessary aspects into their well-defined areas. It just seems we're only gonna get so much mileage with the current approach. 

Here's what I propose: 
```
  // 0. looking up the parent data context with the special "../" syntax
  Component.prototype._parentData(level: Int)

  // 1. look up a helper on the current template (this is what you previously built)
  Component.prototype._getTemplateHelper(name: String)

  // 2. look up a binding by traversing the lexical view hierarchy inside the current template
  Component.prototype._lexicalBindingLookup(name: String);
  //note: I'm pretty sure these are things such as `@index` https://github.com/meteor/meteor/blob/dc3cd6eb92f2bdd1bb44000cdd6abd1e5d0285b1/packages/blaze/builtins.js#L186 , which are very few. I think Blaze also uses it to find the binding of `person` in this example: `{{#each person in people}}`

  // 3. look up a template by name (since they can be used as helpers: `{{#templateName}}`)
  Component._getTemplate(name: String);
   
  // 4. look up a global helper
  Component._getGlobalHelper(name: String);

  // 5. look up in a data context (this is basically what you started today and yesterday)
  Component.prototype.currentData()
```

Now the question after that is does a "dumb" React component actually have all this information. It likely doesn't, which is why you're using workarounds in the replacements to set such things. But I think at the very least having these method names available will do a lot for planning our attack. We can parameterize these methods more so they get more runtime data you transpiled into the render methods via replacements. 

`../..` might really be a problem. It violates the principles of React. It's the exact opposite of how you pass down props in react. It's a template instance reaching backward. I wish we could statically address this, but you can't since a template instance can be included by multiple parents. We got 2 options: A) pass along parent context and combine with childrens, overwriting similar prop names by parent if necessary, until we have a huge data context available to us that is the combination of all previous lineage. B) in every component instance we have essentially a linked list of parent references available for us to traverse. The latter is obviously the better solution. But we may even be better off not supporting this feature, and requiring developers write this aspect of their Blaze code so that they pass down such context to be used with out `../`. Here is some quick research on how there essentially isn't a good solution for us in React already:

 - https://facebook.github.io/react/tips/expose-component-functions.html
 - http://stackoverflow.com/questions/29051628/react-whats-the-right-way-to-get-a-parent-props-from-his- - children
https://www.tildedave.com/2014/11/15/introduction-to-contexts-in-react-js.html (i just did a quick peak, but my gut says its not the way to go; I could be wrong though)

Our overall goal should be so that you can switch out your renderer and your Blaze code will just work with a React renderer. But I'd say there is a bunch of lesser used features that--if they cause us major problems--we should not support, and rather document. Perhaps in the future we can even provide a converter script that converts their code for them. I do think we absolutely need to transpilate `Template.instance().foo = new ReactiveVar/Dict/Field` into react style `state`, but then not do things like `../..` since it violates the very principles of React, and will likely lead to major problems. ..But I also could be completely wrong--React may just not have constructs for it, but if a parent has the context, the child will have the same unchanged context. It shouldn't violate its inputs. It's just gonna be a lot of extra props being passed down. What I'm saying is it's definitely possible for us to do--but if it kills performance or completely breaks React, it's a no go. Unfortunately for some of these, we may not even know until we try it. Passing down context for events is essentially the same thing and that must be done. 

In conclusion, #1 and #5 above are top priority; and then #2. If we can build them with a consistent interface, it will pave the way for any others we wanna add. Then I think we do the events stuff so all context is combined and passed down. Once we get that to work, then we come back here and see if we wanna do the same for `../../etc`.

**Comment by faceyspacey** [comment]
i stand corrected--React "Contexts" is actually what we should use:

https://www.tildedave.com/2014/11/15/introduction-to-contexts-in-react-js.html

particularly the `childContextTypes:` version as opposed to `withContext`.

**Comment by mitar** [comment]
> And here are some modifications @mitar made for his Blaze Components package:
> https://github.com/peerlibrary/meteor-blaze-components/blob/master/lookup.js#L94

BTW, this file is not necessary. This was all pulled into Meteor with 1.2. So this file is just for backwards compatibility.

**Comment by mitar** [comment]
Just to pitch few cents here. So the issue with needing global contexts like what Blaze has is I think the thing which is problematic to provide. Blaze Components improved API here and all calls are made against template/component instance, so it is easier to control and easier to make sure that things work consistently across the board.

So yea, trying to support Blaze 1 makes it quite a bit harder for you. :-) Blaze Components API would be easier to do.

**Comment by faceyspacey** [comment]
yea, that's something I've been wrestling with. But even tho I also have a Blaze Components package, I think it would be really nice for the community to be able to switch out the renderer to React for the original Blaze. It's definitely harder, but it does seem doable. And even if it isn't the most performant use of React, it's a good step in that direction, and they can optimize later. But more importantly it opens the doors for React Native and the rest of the React ecosystem. Just imagine Blaze 1.0 rendered reactively natively. 

ps. Once we get Blaze 1.0 to work, it will be very quick to make it work for both our component systems.

**Comment by mitar** [comment]
I agree. This is why I think that for now we should try to mirror Blaze 1, with all the quirks it has. Yes.

**Comment by timbrandin** [comment]
@faceyspacey I already have a solution for #2 in your comment above if we talk about `person` in persons. See https://github.com/timbrandin/react-templates/blob/master/exports.js#L31-L40

**Comment by faceyspacey** [comment]
yea i saw that, i just wasnt sure if it captured all cases. it seems that is #2 and #5 wrapped in one. i.e. both `@index` and the `let` variable name (which Blaze considers "bindings") and #5, which is data contexts. 

It likely doesnt work yet with custom block helpers: https://github.com/meteor/meteor/blob/devel/packages/spacebars/README.md#custom-block-helpers , but that doesnt matter yet obviously. 

...I guess the main thing in my mind was: does this solution truly allow us to traverse the data context chain backwards. Is `__context` linked to the previous `context` and so on, or does the chain get broken?? I think that's the idea, but I can't quite tell for sure.

I guess "parentData" doesn't matter at all in templates. As long as `#with/each` keep setting a new scoped value for `context` we are in good shape. The only thing is if we are only using closures, it doesn't allow us to climb back up the scope if we ever wanna make `parentData(level)` possible in helpers/events. I know this is essentially making the current stage unnecessarily complex--we can always come back to it once the base house of cards stands--but I just wanted to bring it up.
